sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample08_EventProcessor.md

```diff
@@ -0,0 +1,230 @@
+        {
+            if (long.TryParse(item.Metadata[OffsetMetadataKey], NumberStyles.Integer, CultureInfo.InvariantCulture, out long offset))
+            {
+                checkpoints.Add(new EventProcessorCheckpoint()
+                {
+                    ConsumerGroup = ConsumerGroup,
+                    EventHubName = EventHubName,
+                    FullyQualifiedNamespace = FullyQualifiedNamespace,
+                    PartitionId = item.Name.Substring(checkpointBlobsPrefix.Length),
+                    StartingPosition = EventPosition.FromOffset(offset, isInclusive: false)
+                });
+            }
+        }
+
+        return checkpoints;
+    }
+
+    // Allow subclasses to call CheckpointAsync to store checkpoint information without having to understand the details of how checkpoints are stored.
+    protected async Task CheckpointAsync(TPartition partition, EventData data, CancellationToken cancellationToken = default)
+    {
+        string checkpointBlob = string.Format(CheckpointPrefixFormat + partition.PartitionId, FullyQualifiedNamespace.ToLowerInvariant(), EventHubName.ToLowerInvariant(), ConsumerGroup.ToLowerInvariant());
+        Dictionary<string, string> checkpointMetadata = new Dictionary<string, string>()
+        {
+            { OffsetMetadataKey, data.Offset.ToString(CultureInfo.InvariantCulture) },
+        };
+
+        using MemoryStream emptyStream = new MemoryStream(Array.Empty<byte>());
+        await StorageContainer.GetBlobClient(checkpointBlob).UploadAsync(emptyStream, metadata: checkpointMetadata, cancellationToken: cancellationToken).ConfigureAwait(false);
+    }        
+}
+```
+
+## Using the Custom Processor
+
+The `AzureBlobStorageEventProcessor<TPartiton>` above is abstract because we did not implement any of the `On*` methods. By separating the logic for state management and event processing, we can leverage `AzureBlobStorageEventProcessor<TPartiton>` to build many different processors, each with their own business logic, by extending `AzureBlobStorageEventProcessor<TPartiton>` and implementing `OnProcessingEventBatchAsync` and `OnProcessingErrorAsync`.
+
+Here's a simple processor that just counts the number of events received in a batch and then creates a checkpoint so they are not processed again:
+
+```C# Snippet:EventHubs_Sample08_CustomProcessor
+class CustomProcessor : AzureBlobStorageEventProcessor<EventProcessorPartition>
+{
+    public CustomProcessor(int eventBatchMaximumCount, string consumerGroup, string fullyQualifiedNamespace, string eventHubName, TokenCredential credential, BlobContainerClient storageContainer, EventProcessorOptions options = null) : base(eventBatchMaximumCount, consumerGroup, fullyQualifiedNamespace, eventHubName, credential, storageContainer, options)
+    {
+    }
+
+    protected async override Task OnProcessingEventBatchAsync(IEnumerable<EventData> events, EventProcessorPartition partition, CancellationToken cancellationToken)
```

Comment:
You may want to mention that this method will be invoked concurrently so the implementation needs to be thread-safe and that there will only be one active invocation for any partition to ensure that the events from the partition arrive in order.

Reply:
Maybe also that there will be one instance of `TPartition` created per partition and managed by the processor.  That instance will be passed each time for the partition, so the handler methods can count on a consistent instance.   If you're trying to preserve application state, the `TPartition` instance is a good place to do so.